firedrake/pointwise_operators.py

```diff
@@ -0,0 +1,813 @@
+        :param disp: boolean indication whether we display the max of the number of iterations taken over the pointwise solves.
+        """
+
+        AbstractPointwiseOperator.__init__(self, *operands, function_space=function_space, derivatives=derivatives, count=count, val=val, name=name, dtype=dtype, operator_data=operator_data)
+
+        # Check
+        if not isinstance(operator_data['point_solve'], types.FunctionType):
+            error("Expecting a FunctionType pointwise expression")
+
+        if operator_data['point_solve'].__code__.co_argcount != len(operands) + 1:
+            error("Expecting %s operands" % (operator_data['point_solve'].__code__.co_argcount-1))
+        if operator_data['solver_name'] not in ('newton', 'secant', 'halley'):
+            error("Expecting of the following method : %s" % ('newton', 'secant', 'halley'))
+        if not isinstance(operator_data['solver_params'], dict):
+            error("Expecting a dict with the solver arguments instead of %s" % operator_data['solver_params'])
+
+        self.disp = disp
+
+    @property
+    def operator_f(self):
+        return self.operator_data['point_solve']
+
+    @property
+    def solver_name(self):
+        return self.operator_data['solver_name']
+
+    @property
+    def solver_params(self):
+        return self.operator_data['solver_params']
+
+    def compute_derivatives(self, f):
+        deriv_index = (0,) + self.derivatives
+        symb = (self._sympy_create_symbols(self.ufl_function_space().shape, 0),)
+        symb += tuple(self._sympy_create_symbols(e.ufl_shape, i+1) for i, e in enumerate(self.ufl_operands))
+        ufl_space = self.ufl_function_space()
+        ufl_shape = self.ufl_shape
+
+        if deriv_index == (0,)*len(deriv_index):
+            return f
+
+        fexpr = self.operator_f(*symb)
+        args = tuple(Function(ufl_space).interpolate(pi) for pi in self.ufl_operands)
+        vals = tuple(coeff.dat.data_ro for coeff in args)
+        for i, di in enumerate(deriv_index):
+            if di != 0:
+                def implicit_differentiation(res):
+                    # We want to compute dsymb[0]/dsymb[i]
+                    # We know by Implicit Function Theorema that :
+                    # if f(x,y) = 0 =>  df/dx + dy/dx*df/dy = 0
+                    dfds0 = sp.diff(fexpr, symb[0])
+                    dfdsi = sp.diff(fexpr, symb[i])
+
+                    # -> Sympy idiff does not work for high dimension
+                    # -> We could solve the system "dfds0*ds0/dsi = -dfdsi" symbolically and then
+                    # for the n-th derivative just apply n-1th time the derivative on the result.
+                    # However again sympy.linsolve() does not handle high dimensional case
+                    dfds0l = sp.lambdify(symb, dfds0, modules='numpy')
+                    dfdsil = sp.lambdify(symb, dfdsi, modules='numpy')
```

Comment:
Please add `dummify=True` at the very least. Question, should we sanitise input, since this just calls `exec`

Reply:
Ok thank you. I don't really know, I would say yes if we can.

Because `fexpr` is by assumption supposed to be a SymPy expression passed in the `point_solve` helper function by the user himself, then we just compose it with other SymPy operations (`sp.diff` and `sp.lambdify`). So any problems with the input should come from an incorrect SymPy expression passed in the `point_solve`, which the user can correct (I think).